### DISCUSSION TITLE
How can I reliably retrieve the weight classification of a font (Condensed, Compressed, Expanded, Wide, Etc.) using FontTools / TTX / FontForge CLI?

### DISCUSSION THREAD

**Discussion by futuremotiondev** [fonttools/fonttools#3157] [Q&A]
I am writing a script that aims to reliably sort a font family into it's individual weights if present. For instance:

```
SomeFont Pro.otf
SomeFont Pro Black.otf
SomeFont Pro Black Italic.otf
SomeFont Pro Bold.otf
SomeFont Pro Bold Italic.otf
SomeFont Pro Condensed.otf
SomeFont Pro Condensed Black.otf
SomeFont Pro Condensed Black Italic.otf
SomeFont Pro Condensed Bold.otf
SomeFont Pro Condensed Bold Italic.otf
SomeFont Pro Condensed Italic.otf
SomeFont Pro Condensed Light.otf
SomeFont Pro Condensed Light Italic.otf
SomeFont Pro Condensed Medium.otf
SomeFont Pro Condensed Medium Italic.otf
SomeFont Pro Condensed Semibold.otf
SomeFont Pro Condensed Semibold Italic.otf
SomeFont Pro Extra Light.otf
SomeFont Pro Extra Light Italic.otf
SomeFont Pro ExtraCondensed.otf
SomeFont Pro ExtraCondensed Black.otf
SomeFont Pro ExtraCondensed Black Italic.otf
SomeFont Pro ExtraCondensed Bold.otf
SomeFont Pro ExtraCondensed Bold Italic.otf
SomeFont Pro ExtraCondensed Italic.otf
SomeFont Pro ExtraCondensed Light.otf
SomeFont Pro ExtraCondensed Light Italic.otf
SomeFont Pro ExtraCondensed Med.otf
SomeFont Pro ExtraCondensed Med Italic.otf
SomeFont Pro Italic.otf
SomeFont Pro Light.otf
SomeFont Pro Light Italic.otf
SomeFont Pro Medium.otf
SomeFont Pro Medium Italic.otf
SomeFont Pro Semibold.otf
SomeFont Pro Semibold Italic.otf
SomeFont Pro SemiCondensed.otf
SomeFont Pro SemiCondensed Black.otf
SomeFont Pro SemiCondensed Black Italic.otf
SomeFont Pro SemiCondensed Bold.otf
SomeFont Pro SemiCondensed Bold Italic.otf
SomeFont Pro SemiCondensed ExtLt.otf
SomeFont Pro SemiCondensed ExtLt Italic.otf
SomeFont Pro SemiCondensed Italic.otf
SomeFont Pro SemiCondensed Light.otf
SomeFont Pro SemiCondensed Light Italic.otf
SomeFont Pro SemiCondensed Medium.otf
SomeFont Pro SemiCondensed Medium Italic.otf
SomeFont Pro Thin.otf
SomeFont Pro Thin Italic.otf
SomeFont Pro Ultra Black.otf
SomeFont Pro Ultra Black Italic.otf
SomeFont Pro Wide.otf
SomeFont Pro Wide Black.otf
SomeFont Pro Wide Black Italic.otf
SomeFont Pro Wide Bold.otf
SomeFont Pro Wide Bold Italic.otf
SomeFont Pro Wide Italic.otf
SomeFont Pro Wide Light.otf
SomeFont Pro Wide Light Italic.otf
SomeFont Pro Wide Medium.otf
SomeFont Pro Wide Medium Italic.otf
SomeFont Pro Wide Semibold.otf
SomeFont Pro Wide Semibold Italic.otf
```

After the script runs, it should create the following directories: `Condensed`, `ExtraCondensed`, `SemiCondensed`, `Wide` - and then move the appropriate fonts to these folders. 

The small dataset above is just for illustration, obviously I could just search for these keywords in the filenames but this isn't at all reliable as many fonts are named more cryptically or follow different naming conventions.

I've got a Python VENV set up with `FontTools`, `TTX`, and `ftcli` set up in an attempt to extract metadata so I can make more informed decisions in my script. 

I have a script that uses `ttLib` to extract the `OS/2` [usWidthClass][1]:

```py
import os
import sys
import pathlib
from contextlib import redirect_stderr
from fontTools import ttLib

if len(sys.argv) > 2:
    print("Usage: Only one argument required: [FONT]")
    exit(1)

fontfile = sys.argv[1]
font = ttLib.TTFont(fontfile)
width = font['OS/2'].usWidthClass

print(width)
```

And a script to dump metadata using `FontMeta`:

```py
import os.path
import sys
from fontTools import ttLib
from fontmeta import FontMeta

if len(sys.argv) > 1:
    print("Usage: Only one argument required: [FONT]")
    exit(1)

fontfile = sys.argv[1]
meta_instance = FontMeta(fontfile)
metadata = meta_instance.get_full_data()

print(metadata)
```

But I've found the usWidthClass to be hit or miss. A lot of typographers just set this value as `5` (Normal) no matter what width the font is. And the data extracted from FontMeta doesn't really supply any useful information for this problem.

I could really use some guidance on how to move forward and scrape the information I need. Right now I'm just doing regex gymnastics on the supplied filenames and suspect there is a much more reliable way to go about this.

Any help greatly appreciated.

  [1]: https://learn.microsoft.com/en-us/typography/opentype/spec/os2#uswidthclass

**Comment by ftCLI** [comment]
What you are trying to do, If I well understood, is sorting fonts in folders. Since you have already installed ftCLI, you can just use `ftcli utils font-organizer INPUT_PATH`.

https://github.com/ftCLI#ftcli-utils-font-organizer

If the fonts are not grouped as a superfamily (width style is in Family Name, not in Subfamily Name), it will do exactly what asked.

Otherwise, the question is a bit more complex.

**Reply by futuremotiondev** [reply]
I am indeed looking to sort fonts in folders, but in a very specific way. I only want to create sub-folders for font width. And I want to keep the width naming consistent.  I don't see an option with font-organizer to sort fonts by width unfortunately.

```
--rename-source [1|2|3|4|5]  Renames the font files according to the
                             provided source string(s). See ftcli utils
                             font-renamer.
-ext, --extension            Sorts fonts by extension.
-ver, --version              Sorts fonts by version.
```

**Reply by ftCLI** [reply]
Yes, that works only if fonts with different widthClass have different family name.

```
nameID 16: Some Font Pro Wide
nameID 17: Black Italic
```

 Running `ftcli utils font-organizer` will **implicitly** sort fonts by width and move the file to a directory named "Some Font Pro Wide". This won't happen in the following situation:

```
nameID 16: Some Font Pro
nameID 17: Wide Black Italic
```
 
If you really want to sort fonts to in folders named after width style name, you should create a dict and map usWidthClass values to style names:

```python
import os

import click
from fontTools.misc.cliTools import makeOutputFileName

from ftCLI.Lib.Font import Font
from ftCLI.Lib.utils.cli_tools import check_input_path
from ftCLI.Lib.utils.click_tools import add_file_or_path_argument, generic_info_message, generic_error_message

WIDTHS = {
    1: "Compressed",
    2: "ExtraCondensed",
    3: "Condensed",
    4: "Narrow",
    5: "Normal",
    6: "Wide",
    7: "Extended",
    8: "ExtraExtended",
    9: "Expanded"
}


@click.command()
@add_file_or_path_argument()
def cli(input_path):

    files = check_input_path(input_path)

    for file in files:
        try:
            font = Font(file)
            width_class = font.os_2_table.usWidthClass
            width_style_name = WIDTHS.get(width_class)

            file_name = os.path.basename(file)
            output_dir = os.path.join(os.path.dirname(file), width_style_name)

            output_file = makeOutputFileName(file_name, outputDir=output_dir, overWrite=False)
            os.renames(file, output_file)
            generic_info_message(f"{os.path.basename(file)} moved")

        except Exception as e:
            generic_error_message(e)

if __name__ == "__main__":
    cli()
```

> But I've found the usWidthClass to be hit or miss. A lot of typographers just set this value as 5 (Normal) no matter what width the font is. And the data extracted from FontMeta doesn't really supply any useful information for this problem.

This is a bit tricky: if you don't trust usWidthClass and want to recalculate it from a string, you can give a look here and adapt the function to your needs: https://github.com/ftCLI/ftCLI/blob/98e9a489e95f4486159081f971d5807fda0ba8b3/ftCLI/Lib/assistant/fonts_data.py#L31
